packages/kbn-text-based-editor/src/text_based_languages_editor.tsx

```diff
@@ -394,15 +435,52 @@ export const TextBasedLanguagesEditor = memo(function TextBasedLanguagesEditor({
     ]
   );
 
+  const parseMessages = useCallback(async () => {
+    if (editorModel.current) {
+      return await ESQLLang.validate(editorModel.current, queryString, esqlCallbacks);
+    }
+    return {
+      errors: [],
+      warnings: [],
+    };
+  }, [esqlCallbacks, queryString]);
+
+  useEffect(() => {
+    const validateQuery = async () => {
+      if (editorModel?.current) {
+        const parserMessages = await parseMessages();
+        setClientParserMessages({
+          errors: parserMessages?.errors ?? [],
+          warnings: parserMessages?.warnings ?? [],
+        });
+      }
+    };
+    if (isQueryLoading || isLoading) {
+      addQueriesToCache({
+        queryString,
+        timeZone,
+      });
+      validateQuery();
+      setRefetchHistoryItems(false);
+    } else {
+      updateCachedQueries({
+        queryString,
+        status: clientParserMessages.errors?.length
+          ? 'error'
+          : clientParserMessages.warnings.length
+          ? 'warning'
+          : 'success',
+      });
+
+      setRefetchHistoryItems(true);
```

Comment:
Do I understand correctly that `refetchHistoryItems` is essentially a bouncer?

If so, I'm wondering about removing [this check](https://github.com/elastic/kibana/pull/178302/files#diff-7eb917ee0c77516d0cda2258336c062288a5bdee27764f1d089a0abe12ff4865R204) and defining refetchHistoryItems as a function in this file?

```ts
const refetchHistoryItems = useCallback(() => setRefetchHistoryBouncer(!refetchHistoryBouncer));
```

Then, `refetchHistoryItems` can just be called in the logic here whenever a refetch should happen.

Not sure if it's worth it, but it could be an understandability improvement. No strong feelings.

Reply:
I am sorry Drew I can't get your proposal. Can you elaborate more?

Reply:
Am I correct that we're trying to tell a child component to refetch the query list from local storage?

If so, my suggestion was that simply calling a function (`refetchHistoryItems()` better matches this use case). So the logic in this use effect is currently

```ts
  useEffect(() => {
    const validateQuery = async () => {
      ...
    };
    if (isQueryLoading || isLoading) {
      addQueriesToCache({
          ...
      });
      validateQuery();
      setRefetchHistoryItems(false);
    } else {
      updateCachedQueries({
          ...
      });

      setRefetchHistoryItems(true);
    }
  }, [...]);
```
But it could be

```ts
  useEffect(() => {
    const validateQuery = async () => {
      ...
    };
    if (isQueryLoading || isLoading) {
      addQueriesToCache({
          ...
      });
      validateQuery();
    } else {
      updateCachedQueries({
          ...
      });

      refetchHistoryItems();
    }
  }, [...]);
```

How the function accomplishes this is an implementation detail. In this case, the function could toggle a flag "behind the scenes."

```ts
const refetchHistoryItems = useCallback(() => setRefetchHistoryBouncer(!refetchHistoryBouncer));
```

The more I explain this, the more unimportant and nitpicky it feels. Feel free to ignore :D